src/subshader/dsp/wavelet.py

```diff
@@ -1,464 +1,643 @@
+    def __init__(self,
+                 sample_rate: int,
+                 input_n: int,
+                 config: Optional[WaveletConfig] = None) -> None:
+        """
+        CWT implementation from Analyzing Neural Time Series (ANTS) by Mike X 
+        Cohen translated from Matlab. According to the list of frequencies, we
+        generate a bank of wavelet kernels on init. Then determine the output's
+        reliable region based on the wavelet with the widest time support. All
+        of this in preparation for the CWT call, which performs the convolution
+        between the audio data and the bank of wavelet kernels.
 
         Args:
-            sample_rate (int): The rate the data was sampled in Hz
-            window_size (int): The length of the data
+            sample_rate: Input audio sample rate in Hz.
+            input_n: Input data length in samples.
+            num_cycles: Number of carrier cycles per wavelet.
+            fwhm_cycles: Gaussian FWHM width in cycles.
+            config: Configuration object with wavelet parameters.
         """
-        # TODO ISSUE-36 Use class inheritance effectively to avoid code 
-        # duplication
-        super().__init__(sample_rate, window_size)
+        super().__init__(sample_rate, input_n, config)
 
-        # Initialize the time-frequency matrix
-        self.tf = np.zeros((self.num_freqs, self.window_size))
+        self.wavelets: list[WaveletKernel] = [WaveletKernel(f=f, 
+                                                            sample_rate=self.sample_rate, 
+                                                            num_cycles=self.config.num_cycles, 
+                                                            num_fwhm_cycles=self.config.num_fwhm_cycles, 
+                                                            input_n=self.input_n) for f in self.freqs]
 
-        # Create a centered time vector for the CMW
-        cmw_t = np.arange(2*self.sample_rate) / self.sample_rate
-        self.cmw_t = cmw_t - np.mean(cmw_t) 
+        self.num_wavelets: int = len(self.wavelets)
 
-        # N's of convolution
-        self.data_n = self.window_size
-        self.kern_n = len(cmw_t)
-        self.conv_n = self.data_n + self.kern_n - 1
-        self.half_kern_n = self.kern_n // 2
+        # Assess the reliable regions of the CWT output
+        self.reliable_slice: slice = self._create_reliable_slice(self.wavelets)
+
+    def normalize_by_scale(self, cwt_coefs: np.ndarray[np.complexfloating]) -> np.ndarray[np.complexfloating]:
+        """
+        Scale-Dependent Normalization to account for the energy bias introduced
+        by scaling each wavelet. At higher scales (lower frequencies), the 
+        wavelet physically gets wider so naturally it collects more energy. To 
+        compensate for that, we reduce the energy of the cwt's result by square 
+        root of the scale where s ≈ 1/f -> 1/sqrt(s) ≈ sqrt(f) 
```

Comment:
explain the square root part
```suggestion
        root of the scale where s ≈ 1/f -> 1/sqrt(s) ≈ sqrt(f) 
        
        TODO-37 explain the square root is because power is mag^2
```